framework/include/materials/Material.h

```diff
@@ -258,6 +259,25 @@ class Material :
    */
   virtual QpData * createData();
 
+  /**
+   * Get a apecial MaterialUserObject which can be explicitly called by the material
+   * during the material property calculation.
+   * @param name The name of the parameter key of the user object to retrieve
+   * @return The user object with name associated with the parameter 'name'
+   */
+  template<class T>
+  const T & getMaterialUserObject(const std::string & name);
```

Comment:
I don't think these new functions are necessary.  This should just be handled on the back end in the Warehouse.  The Warehouse should return either `tid=0` or the correct `tid` depending on what kind of UserObject is being requested.  `UserObjectInterface` can be modified to pass the `tid` down when it makes the call.

No reason to add extra "special" functions for this new type of `UserObject`